### PR TITLE
sync_diff_inspector: add auto-position of tidb_snapshot

### LINKED ISSUE #663: sync-diff-inspector: auto position tidb_snapshot from last syncpoint
## Feature Request

**Is your feature request related to a problem? Please describe:**

Currently sync-diff-inspector does not support checking streams that are actively being read/written from. This is a problem, since it can't be used for TiCDC consistency checking. There is a FR to do online checksumming here: https://github.com/pingcap/dm/issues/1097

However, if the source and destination are TiDB this is not the best way to do this. Instead, syncpoint (experimental) can be used. The source and destination can be set to use a tidb_snapshot to consistently compare.

**Describe the feature you'd like:**

My feature request is to "auto position" the `snapshot` which is used to compare, based on reading the most recently entry from the syncpoint table.

For example, given the following config:

```toml
######################### Databases config #########################
[data-sources]
[data-sources.tidb1]
    host = "127.0.0.1"
    port = 4000
    user = "root"
    password = ""
    snapshot = "auto"

[data-sources.tidb2]
    host = "192.168.86.38"
    port = 4000
    user = "root"
    password = ""
    snapshot = "auto"

######################### Task config #########################
# Required
[task]
    # 1 fix sql: fix-target-TIDB1.sql
    # 2 log: sync-diff.log
    # 3 summary: summary.txt
    # 4 checkpoint: a dir
    output-dir = "/tmp/output/config"

    source-instances = ["tidb1"]

    target-instance = "tidb2"

    # tables need to check. *Include `schema` and `table`. Use `.` to split*
    target-check-tables = ["test.stock"]

    # extra table config
    target-configs= ["config1"]
```

sync-diff-inspector will automatically run the following query on `target-instance` and apply primary_ts to source and secondary_ts to target.

```sql
tidb> select primary_ts, secondary_ts  from tidb_cdc.syncpoint_v1 order by primary_ts desc limit 1;
+--------------------+--------------------+
| primary_ts         | secondary_ts       |
+--------------------+--------------------+
| 434722879294406656 | 434723226951352321 |
+--------------------+--------------------+
1 row in set (0.00 sec)
```

This will save a lot of config file editing.

**Describe alternatives you've considered:**

The alternative is to generate config files each time to use for compare.  But it's a pretty straight forward use case so I'm hoping support can be built in.

**Teachability, Documentation, Adoption, Migration Strategy:**

### PR DESCRIPTION
**Pull request by morgo**
### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tidb-tools/issues/663

### What is changed and how it works?

This supports an `auto` value for snapshot. It must be set on both source and target, and only supports one source.

When set, sync_diff_inspector will read the syncpoint table for the target and auto-set the `tidb_snapshot` value.

### Check List 

Tests 

 - Unit test
 - Integration test
 - Manual test (add detailed scripts or steps below)

First when the syncpoint table doesn't exist, the error message is quite readable:
```
[2022/07/20 14:11:43.672 -06:00] [FATAL] [main.go:120] ["failed to initialize diff process"] [error="fetching auto-position tidb_snapshot failed: Error 1146: Table 'tidb_cdc.syncpoint_v1' doesn't exist"] [errorVerbose="Error 1146: Table 'tidb_cdc.syncpoint_v1' doesn't exist\nfetching auto-position tidb_snapshot failed\ngithub.com/pingcap/tidb-tools/sync_diff_inspector/source.getAutoSnapshotPosition\n\t/Users/mtocker/go/src/github.com/pingcap/tidb-tools/sync_diff_inspector/source/source.go:241\ngithub.com/pingcap/tidb-tools/sync_diff_inspector/source.initDBConn\n\t/Users/mtocker/go/src/github.com/pingcap/tidb-tools/sync_diff_inspector/source/source.go:263\ngithub.com/pingcap/tidb-tools/sync_diff_inspector/source.NewSources\n\t/Users/mtocker/go/src/github.com/pingcap/tidb-tools/sync_diff_inspector/source/source.go:116\nmain.(*Diff).init\n\t/Users/mtocker/go/src/github.com/pingcap/tidb-tools/sync_diff_inspector/diff.go:139\nmain.NewDiff\n\t/Users/mtocker/go/src/github.com/pingcap/tidb-tools/sync_diff_inspector/diff.go:97\nmain.checkSyncState\n\t/Users/mtocker/go/src/github.com/pingcap/tidb-tools/sync_diff_inspector/main.go:117\nmain.main\n\t/Users/mtocker/go/src/github.com/pingcap/tidb-tools/sync_diff_inspector/main.go:104\nruntime.main\n\t/usr/local/go/src/runtime/proc.go:250\nruntime.goexit\n\t/usr/local/go/src/runtime/asm_arm64.s:1263"] [stack="main.checkSyncState\n\t/Users/mtocker/go/src/github.com/pingcap/tidb-tools/sync_diff_inspector/main.go:120\nmain.main\n\t/Users/mtocker/go/src/github.com/pingcap/tidb-tools/sync_diff_inspector/main.go:104\nruntime.main\n\t/usr/local/go/src/runtime/proc.go:250"]
```

I then used the following config:

```
######################### Databases config #########################
[data-sources]
[data-sources.tidb1]
    host = "127.0.0.1"
    port = 4000
    user = "root"
    password = ""
    snapshot = "auto"

[data-sources.tidb2]
    host = "192.168.86.38"
    port = 4000
    user = "root"
    password = ""
    snapshot = "auto"

######################### Task config #########################
# Required
[task]
    # 1 fix sql: fix-target-TIDB1.sql
    # 2 log: sync-diff.log
    # 3 summary: summary.txt
    # 4 checkpoint: a dir
    output-dir = "/tmp/output/config"

    source-instances = ["tidb1"]

    target-instance = "tidb2"

    # tables need to check. *Include `schema` and `table`. Use `.` to split*
    target-check-tables = ["test.stock"]

    # extra table config
    target-configs= ["config1"]
```

I then started a benchmark load so data would be replicating via TiCDC:
```
tiup bench tpcc prepare --warehouses 500 --threads 1
```

My first attempt at running it failed (but the error is expected - the tidb_snapshot is too old):

```
$  rm -rf /tmp/output/config && ./bin/sync_diff_inspector --config=sync_diff_inspector/config/config.toml
There is something error when initialize diff, please check log info in /tmp/output/config/sync_diff.log
~/go/src/github.com/pingcap/tidb-tools auto-position-snapshot * cat  /tmp/output/config/sync_diff.log
[2022/07/20 14:22:03.169 -06:00] [INFO] [printer.go:46] ["Welcome to sync_diff_inspector"] ["Release Version"=v5.3.2-dirty] ["Git Commit Hash"=cf6b7a8ae4b5849a6155634352b3712059defb48] ["Git Branch"=auto-position-snapshot] ["UTC Build Time"="2022-07-20 08:21:41"] ["Go Version"=go1.18.3]
[2022/07/20 14:22:03.170 -06:00] [INFO] [main.go:101] [config="{\"check-thread-count\":4,\"export-fix-sql\":true,\"check-struct-only\":false,\"dm-addr\":\"\",\"dm-task\":\"\",\"data-sources\":{\"tidb1\":{\"host\":\"127.0.0.1\",\"port\":4000,\"user\":\"root\",\"password\":\"\",\"sql-mode\":\"\",\"snapshot\":\"auto\",\"route-rules\":null,\"Router\":{\"Selector\":{}},\"Conn\":null},\"tidb2\":{\"host\":\"192.168.86.38\",\"port\":4000,\"user\":\"root\",\"password\":\"\",\"sql-mode\":\"\",\"snapshot\":\"auto\",\"route-rules\":null,\"Router\":{\"Selector\":{}},\"Conn\":null}},\"routes\":null,\"table-configs\":{\"config1\":{\"target-tables\":[\"schema*.table*\",\"test2.t2\"],\"Schema\":\"\",\"Table\":\"\",\"ConfigIndex\":0,\"HasMatched\":false,\"IgnoreColumns\":[\"\",\"\"],\"Fields\":[\"\"],\"Range\":\"age \\u003e 10 AND age \\u003c 20\",\"TargetTableInfo\":null,\"Collation\":\"\",\"chunk-size\":0}},\"task\":{\"source-instances\":[\"tidb1\"],\"source-routes\":null,\"target-instance\":\"tidb2\",\"target-check-tables\":[\"test.stock\"],\"target-configs\":[\"config1\"],\"output-dir\":\"/tmp/output/config\",\"SourceInstances\":[{\"host\":\"127.0.0.1\",\"port\":4000,\"user\":\"root\",\"password\":\"\",\"sql-mode\":\"\",\"snapshot\":\"auto\",\"route-rules\":null,\"Router\":{\"Selector\":{}},\"Conn\":null}],\"TargetInstance\":{\"host\":\"192.168.86.38\",\"port\":4000,\"user\":\"root\",\"password\":\"\",\"sql-mode\":\"\",\"snapshot\":\"auto\",\"route-rules\":null,\"Router\":{\"Selector\":{}},\"Conn\":null},\"TargetTableConfigs\":[{\"target-tables\":[\"schema*.table*\",\"test2.t2\"],\"Schema\":\"\",\"Table\":\"\",\"ConfigIndex\":0,\"HasMatched\":false,\"IgnoreColumns\":[\"\",\"\"],\"Fields\":[\"\"],\"Range\":\"age \\u003e 10 AND age \\u003c 20\",\"TargetTableInfo\":null,\"Collation\":\"\",\"chunk-size\":0}],\"TargetCheckTables\":[{}],\"FixDir\":\"/tmp/output/config/fix-on-tidb2\",\"CheckpointDir\":\"/tmp/output/config/checkpoint\",\"HashFile\":\"\"},\"ConfigFile\":\"sync_diff_inspector/config/config.toml\",\"PrintVersion\":false}"]
[2022/07/20 14:22:03.176 -06:00] [INFO] [common.go:117] ["create connection with snapshot"] [snapshot=434725891369861121]
[2022/07/20 14:22:03.179 -06:00] [INFO] [common.go:117] ["create connection with snapshot"] [snapshot=434725867736006656]
[2022/07/20 14:22:03.181 -06:00] [FATAL] [main.go:120] ["failed to initialize diff process"] [error="create db connections {\"host\":\"127.0.0.1\",\"port\":4000,\"user\":\"root\",\"schema\":\"\",\"snapshot\":\"434725867736006656\"} error Error 8055: snapshot is older than GC safe point 2022-07-20 14:10:57 -0600 MDT"] [errorVerbose="create db connections {\"host\":\"127.0.0.1\",\"port\":4000,\"user\":\"root\",\"schema\":\"\",\"snapshot\":\"434725867736006656\"} error Error 8055: snapshot is older than GC safe point 2022-07-20 14:10:57 -0600 MDT\ngithub.com/pingcap/tidb-tools/sync_diff_inspector/source/common.CreateDB\n\t/Users/mtocker/go/src/github.com/pingcap/tidb-tools/sync_diff_inspector/source/common/conn.go:28\ngithub.com/pingcap/tidb-tools/sync_diff_inspector/source.initDBConn\n\t/Users/mtocker/go/src/github.com/pingcap/tidb-tools/sync_diff_inspector/source/source.go:283\ngithub.com/pingcap/tidb-tools/sync_diff_inspector/source.NewSources\n\t/Users/mtocker/go/src/github.com/pingcap/tidb-tools/sync_diff_inspector/source/source.go:115\nmain.(*Diff).init\n\t/Users/mtocker/go/src/github.com/pingcap/tidb-tools/sync_diff_inspector/diff.go:139\nmain.NewDiff\n\t/Users/mtocker/go/src/github.com/pingcap/tidb-tools/sync_diff_inspector/diff.go:97\nmain.checkSyncState\n\t/Users/mtocker/go/src/github.com/pingcap/tidb-tools/sync_diff_inspector/main.go:117\nmain.main\n\t/Users/mtocker/go/src/github.com/pingcap/tidb-tools/sync_diff_inspector/main.go:104\nruntime.main\n\t/usr/local/go/src/runtime/proc.go:250\nruntime.goexit\n\t/usr/local/go/src/runtime/asm_arm64.s:1263"] [stack="main.checkSyncState\n\t/Users/mtocker/go/src/github.com/pingcap/tidb-tools/sync_diff_inspector/main.go:120\nmain.main\n\t/Users/mtocker/go/src/github.com/pingcap/tidb-tools/sync_diff_inspector/main.go:104\nruntime.main\n\t/usr/local/go/src/runtime/proc.go:250"]
```

I changed the GC time on source/target:
```
set global tidb_gc_life_time = '24h';
```

It now works as expected:

```sql
* rm -rf /tmp/output/config && ./bin/sync_diff_inspector --config=sync_diff_inspector/config/config.toml
A total of 1 tables need to be compared

Comparing the table structure of ``test`.`stock`` ... equivalent
Comparing the table data of ``test`.`stock`` ... equivalent
_____________________________________________________________________________
Progress [============================================================>] 100% 0/0
A total of 1 table have been compared and all are equal.
You can view the comparision details through '/tmp/output/config/sync_diff.log
```

From the log you can see the snapshot values which were used:

```
cat /tmp/output/config/sync_diff.log | grep snapshot
[2022/07/20 14:25:10.507 -06:00] [INFO] [printer.go:46] ["Welcome to sync_diff_inspector"] ["Release Version"=v5.3.2-dirty] ["Git Commit Hash"=cf6b7a8ae4b5849a6155634352b3712059defb48] ["Git Branch"=auto-position-snapshot] ["UTC Build Time"="2022-07-20 08:21:41"] ["Go Version"=go1.18.3]
[2022/07/20 14:25:10.508 -06:00] [INFO] [main.go:101] [config="{\"check-thread-count\":4,\"export-fix-sql\":true,\"check-struct-only\":false,\"dm-addr\":\"\",\"dm-task\":\"\",\"data-sources\":{\"tidb1\":{\"host\":\"127.0.0.1\",\"port\":4000,\"user\":\"root\",\"password\":\"\",\"sql-mode\":\"\",\"snapshot\":\"auto\",\"route-rules\":null,\"Router\":{\"Selector\":{}},\"Conn\":null},\"tidb2\":{\"host\":\"192.168.86.38\",\"port\":4000,\"user\":\"root\",\"password\":\"\",\"sql-mode\":\"\",\"snapshot\":\"auto\",\"route-rules\":null,\"Router\":{\"Selector\":{}},\"Conn\":null}},\"routes\":null,\"table-configs\":{\"config1\":{\"target-tables\":[\"schema*.table*\",\"test2.t2\"],\"Schema\":\"\",\"Table\":\"\",\"ConfigIndex\":0,\"HasMatched\":false,\"IgnoreColumns\":[\"\",\"\"],\"Fields\":[\"\"],\"Range\":\"age \\u003e 10 AND age \\u003c 20\",\"TargetTableInfo\":null,\"Collation\":\"\",\"chunk-size\":0}},\"task\":{\"source-instances\":[\"tidb1\"],\"source-routes\":null,\"target-instance\":\"tidb2\",\"target-check-tables\":[\"test.stock\"],\"target-configs\":[\"config1\"],\"output-dir\":\"/tmp/output/config\",\"SourceInstances\":[{\"host\":\"127.0.0.1\",\"port\":4000,\"user\":\"root\",\"password\":\"\",\"sql-mode\":\"\",\"snapshot\":\"auto\",\"route-rules\":null,\"Router\":{\"Selector\":{}},\"Conn\":null}],\"TargetInstance\":{\"host\":\"192.168.86.38\",\"port\":4000,\"user\":\"root\",\"password\":\"\",\"sql-mode\":\"\",\"snapshot\":\"auto\",\"route-rules\":null,\"Router\":{\"Selector\":{}},\"Conn\":null},\"TargetTableConfigs\":[{\"target-tables\":[\"schema*.table*\",\"test2.t2\"],\"Schema\":\"\",\"Table\":\"\",\"ConfigIndex\":0,\"HasMatched\":false,\"IgnoreColumns\":[\"\",\"\"],\"Fields\":[\"\"],\"Range\":\"age \\u003e 10 AND age \\u003c 20\",\"TargetTableInfo\":null,\"Collation\":\"\",\"chunk-size\":0}],\"TargetCheckTables\":[{}],\"FixDir\":\"/tmp/output/config/fix-on-tidb2\",\"CheckpointDir\":\"/tmp/output/config/checkpoint\",\"HashFile\":\"\"},\"ConfigFile\":\"sync_diff_inspector/config/config.toml\",\"PrintVersion\":false}"]
[2022/07/20 14:25:10.523 -06:00] [INFO] [common.go:117] ["create connection with snapshot"] [snapshot=434726129239326722]
[2022/07/20 14:25:10.533 -06:00] [INFO] [common.go:117] ["create connection with snapshot"] [snapshot=434726025022406656]
```

 - No code

Code changes

 - Has exported function/method change

Side effects

 - Increased code complexity

Related changes

 - Need to update the documentation
 - Need to be included in the release note

### CODE REVIEW

File: sync_diff_inspector/config/config.go [hunk 1]
```diff
@@ -97,6 +98,18 @@ type DataSource struct {
 	// SourceType string `toml:"source-type" json:"source-type"`
 }
 
+// IsAutoSnapshot returns true if the tidb_snapshot is expected to automatically
+// be set from the syncpoint from the target TiDB instance.
+func (d *DataSource) IsAutoSnapshot() bool {
+	return strings.EqualFold(d.Snapshot, "auto")
```
Comment (Leavrth) [inline sync_diff_inspector/config/config.go:104]: Perhaps it would be better to use words such as "auto-cdc", because we also have `dm` and `binlog`.

**Review by Leavrth — APPROVED** [review]

File: sync_diff_inspector/source/source.go [hunk 2]
```diff
@@ -45,8 +45,9 @@ const (
 const UnifiedTimeZone string = "+0:00"
 
 const (
-	ShieldDBName    = "_no__exists__db_"
-	ShieldTableName = "_no__exists__table_"
+	ShieldDBName      = "_no__exists__db_"
+	ShieldTableName   = "_no__exists__table_"
+	GetSyncPointQuery = "SELECT primary_ts, secondary_ts FROM tidb_cdc.syncpoint_v1 ORDER BY primary_ts DESC LIMIT 1"
```
Comment (3pointer) [inline sync_diff_inspector/source/source.go:50]: shall we check whether the table exists before we execute the query?

Comment (morgo) [reply sync_diff_inspector/source/source.go:50]: I don't think you need to - the error message is pretty understandable when the query fails (see manual test).

File: sync_diff_inspector/config/config.go (hunk 1 above)
Comment (morgo) [reply sync_diff_inspector/config/config.go:104]: I prefer auto, but yes this does very specifically relate to the CDC use case. The concern I have with `auto-cdc` is two word tokens are more subject to misconfiguration (autocdc, auto_cdc etc).

Comment (Leavrth) [reply sync_diff_inspector/config/config.go:104]: OK. I think it is acceptable to use `auto`, but we also need to update the document and annotation of configuration template at the same time.

I will update the document after the PR is merged. Could you add the annotation to explain the meanings of value `auto` in the follow configuration in this PR?
https://github.com/pingcap/tidb-tools/blob/1d2f00da8c3e934f3c13332cbd3a1981eb8c1887/sync_diff_inspector/config/config.toml#L30-L32

**Comment by zier-one** [comment]
Hi, Could we print a log to show what tso is used by sync-diff?

**Comment by morgo** [comment]
> Hi, Could we print a log to show what tso is used by sync-diff?

It is already logged at info level. See the manual test:

> [2022/07/20 14:22:03.176 -06:00] [INFO] [common.go:117] ["create connection with snapshot"] [snapshot=434725891369861121]
> [2022/07/20 14:22:03.179 -06:00] [INFO] [common.go:117] ["create connection with snapshot"] [snapshot=434725867736006656]

**Review by 3pointer — APPROVED** [review]
LGTM

**Comment by zier-one** [comment]
LGTM

**Pull request merged**